### DISCUSSION TITLE
Unique constraint - Prevent or force the error?

### DISCUSSION THREAD

**Discussion by cehenriques** [prisma/prisma#19251] [Q&A] [answered]
Hi,

I have a model with a unique constraint, the `name` of the object must be unique.
My question is what's the preferred way of dealing with unique constraints when creating new records?

1. Should I do a findUnique() first and then throw an error if another record already exists with the same name?
2. Or should I wrap the create() method on a try/catch and then parse the error object and find the specific error code of this constraint (I believe it's the P2002)?

I think on the first one I'm writing more code and doing an extra SELECT query just to avoid having an error being logged on the server. 
What's your preference?

Thanks.

**Comment by ludralph** [comment, marked as answer]
Hi @cehenriques 👋 

When dealing with unique constraints in creating new records, both approaches you mentioned are valid and have their pros and cons. 

> Using findUnique() before creating:

This approach involves explicitly checking if a record with the same name already exists using a findUnique() or similar method. If a record is found, you can throw an error or handle it accordingly. The advantage of this approach is that it allows you to provide a more specific error message to the user, indicating that the name they entered is already taken. However, as you mentioned, it does require an additional SELECT query to check for the existence of a record, which can introduce some overhead.

> Wrapping create() in a try/catch block:

This approach involves attempting to create the record directly using the create() method and handling any potential errors that arise. If a duplicate name violates the unique constraint, an error (such as P2002) will be thrown by the database. By catching this error, you can parse the error object and determine if it corresponds to the unique constraint violation. The advantage of this approach is that it avoids the additional SELECT query, which can be beneficial in terms of performance. However, the error message returned by the database may not be as user-friendly or specific as the one you could provide by checking with findUnique().

In terms of preference, it depends on your specific use case. If providing a user-friendly error message is crucial, using findUnique() might be a better choice. On the other hand, if performance and minimizing database queries are more important, wrapping create() in a try/catch block can be a suitable option. You could also consider using a combination of both approaches, where you first attempt to create the record and handle any potential errors, and then, if needed, fallback to a more explicit findUnique() check to provide a better user experience.

Ultimately, the decision depends on the trade-offs you are willing to make based on your application's requirements and priorities.

**Reply by ludralph** [reply]
Hi @cehenriques 


Do you have any more questions about this issue? If not, it would be really helpful if you could quickly **mark this Discussion as _answered_** to indicate that your issue has been resolved. 

In case you do have any outstanding questions, feel free to let us know and we'd be happy to help!

**Reply by ludralph** [reply]
Hi @cehenriques 

Do you have any more questions about this issue? If not, it would be really helpful if you could quickly **mark this Discussion as _answered_** to indicate that your issue has been resolved. 

In case you do have any outstanding questions, feel free to let us know and we'd be happy to help!
